File: deck/seed/code/native/kotlin/runtime/bit.kt
// Bitwise integer operations over the JVM (full 64-bit Long). Reached only through the public bit API.
object bit {
    fun and(left: Long, right: Long): Long = left and right
    fun or(left: Long, right: Long): Long = left or right
    fun exclusiveOr(left: Long, right: Long): Long = left xor right
    fun not(value: Long): Long = value.inv()
    fun shiftLeft(value: Long, count: Long): Long = value shl count.toInt()
    fun shiftRight(value: Long, count: Long): Long = value shr count.toInt()
    fun shiftRightUnsigned(value: Long, count: Long): Long = value ushr count.toInt()
}

File: deck/seed/code/native/browser/runtime/bit.ts
// Bitwise integer operations over the Web platform, 64-bit like the compiled backends: each op runs
// through BigInt and truncates back to a signed 64-bit value, so a mask such as 0xffffffff is a
// mask and not the JavaScript 32-bit coercion to -1. The result rides a plain number, so it is
// exact within the 53-bit safe-integer range (the same bound every number on this host lives
// under). Reached only through the public bit API.
const bit = {
  and: (left: number, right: number): number =>
    Number(
      BigInt.asIntN(
        64,
        BigInt(Math.trunc(left)) & BigInt(Math.trunc(right)),
      ),
    ),
  or: (left: number, right: number): number =>
    Number(
      BigInt.asIntN(
        64,
        BigInt(Math.trunc(left)) | BigInt(Math.trunc(right)),
      ),
    ),
  exclusiveOr: (left: number, right: number): number =>
    Number(
      BigInt.asIntN(
        64,
        BigInt(Math.trunc(left)) ^ BigInt(Math.trunc(right)),
      ),
    ),
  not: (value: number): number =>
    Number(BigInt.asIntN(64, ~BigInt(Math.trunc(value)))),
  shiftLeft: (value: number, count: number): number =>
    Number(
      BigInt.asIntN(
        64,
        BigInt(Math.trunc(value)) << BigInt(Math.trunc(count)),
      ),
    ),
  // arithmetic (sign-preserving), matching the compiled backends
  shiftRight: (value: number, count: number): number =>
    Number(
      BigInt.asIntN(
        64,
        BigInt(Math.trunc(value)) >> BigInt(Math.trunc(count)),
      ),
    ),
  shiftRightUnsigned: (value: number, count: number): number =>
    Number(
      BigInt.asUintN(64, BigInt(Math.trunc(value))) >>
        BigInt(Math.trunc(count)),
    ),
}
